File: problems/Sliding Window Median/medianSlidingWindow.cs
public class Solution {
    public double[] MedianSlidingWindow(int[] nums, int k) {
        var result = new List<double>();

        if (k <= 0 || nums.Length == 0) return result.ToArray();

        var slidingWindow = new List<long>();

        for (int i = 0; i < nums.Length; i++) {
            var target = nums[i];

            var left = 0;
            var right = slidingWindow.Count;

            while (left < right) {
                var mid = left + (right - left) / 2;

                if (slidingWindow[mid] < target) {
                    left = mid + 1;
                } else {
                    right = mid;
                }
            }

            slidingWindow.Insert(left, target);

            if (i >= k - 1) {
                if (i >= k) {
                    slidingWindow.Remove(nums[i - k]);
                }

                if (k % 2 == 1) result.Add(slidingWindow[k / 2]);
                else {
                    result.Add((slidingWindow[k / 2 - 1] + slidingWindow[k / 2]) / 2.0);
                }
            }
        }

        return result.ToArray();
    }
}

File: problems/Sliding Window Median/medianSlidingWindow.java
class Solution {
    public double[] medianSlidingWindow(int[] nums, int k) {
        int n = nums.length;
        double[] result = new double[n - k + 1];
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>((a,b)->Integer.compare(b,a));
        HashMap<Integer, Integer> outNumToCountMap = new HashMap<>();

        for (int idx = 0; k > idx; ++idx) {
            maxHeap.add(nums[idx]);
        }
        for (int idx = 0; (k / 2) > idx; ++idx) {
            minHeap.add(maxHeap.poll());
        }

        result[0] = (1 == (k&1)) ? maxHeap.peek() : 0.5*maxHeap.peek()  + 0.5*minHeap.peek();

        for (int idx = 0; n - k > idx; ++idx) {
            int outNum = nums[idx];
            int inNum = nums[idx + k];
            boolean outFromLeft = true;
            boolean inIntoLeft = true;
            
            outNumToCountMap.put(outNum, 1 + outNumToCountMap.getOrDefault(outNum, 0));
            
            if (maxHeap.peek() >= outNum) {
                outFromLeft = true;
            } else {
                outFromLeft = false;
            }

            if (maxHeap.peek() >= inNum) {
                maxHeap.add(inNum);
                inIntoLeft = true;
            } else {
                minHeap.add(inNum);
                inIntoLeft = false;
            }

            if (outFromLeft && !inIntoLeft) {
                maxHeap.add(minHeap.poll());
            } else if (!outFromLeft && inIntoLeft) {
                minHeap.add(maxHeap.poll());
            }
            
            while (0 < outNumToCountMap.getOrDefault(maxHeap.peek(), 0)) {
                outNumToCountMap.put(maxHeap.peek(), outNumToCountMap.get(maxHeap.poll()) - 1);
            }
            while (0 < outNumToCountMap.getOrDefault(minHeap.peek(), 0)) {
                outNumToCountMap.put(minHeap.peek(), outNumToCountMap.get(minHeap.poll()) - 1);
            }

            result[1 + idx] = (1 == (k&1)) ? maxHeap.peek() : 0.5*maxHeap.peek()  + 0.5*minHeap.peek();
        }

        return result;
    }
}
